File: ios/AlarmScheduler/AlarmSchedulerOccurrenceStore.swift
import Foundation

enum AlarmSchedulerOccurrenceStore {
  private static let occurrencesKey = "alarm_scheduler_occurrences"
  private static let resolutionsKey = "alarm_scheduler_occurrence_resolutions"

  static func save(_ occurrence: [String: Any]) {
    guard let occurrenceId = occurrence["occurrenceId"] as? String, !occurrenceId.isEmpty else {
      return
    }
    var occurrences = storedOccurrences()
    occurrences[occurrenceId] = occurrence
    UserDefaults.standard.set(occurrences, forKey: occurrencesKey)
  }

  static func occurrence(id: String) -> [String: Any]? {
    storedOccurrences()[id]
  }

  static func all(alarmId: String? = nil) -> [[String: Any]] {
    storedOccurrences().values
      .filter { occurrence in
        alarmId == nil || occurrence["alarmId"] as? String == alarmId
      }
      .sorted { scheduledFor($0) > scheduledFor($1) }
  }

  static func updatePhase(occurrenceId: String, phase: String) {
    guard var occurrence = occurrence(id: occurrenceId) else {
      return
    }
    occurrence["phase"] = phase
    save(occurrence)
  }

  static func alarmId(for occurrenceId: String) -> String {
    occurrence(id: occurrenceId)?["alarmId"] as? String ?? occurrenceId
  }

  static func resolution(alarmId: String, idempotencyKey: String) -> [String: Any]? {
    storedResolutions()[resolutionKey(alarmId: alarmId, idempotencyKey: idempotencyKey)]
  }

  static func saveResolution(alarmId: String, idempotencyKey: String, result: [String: Any]) {
    var resolutions = storedResolutions()
    resolutions[resolutionKey(alarmId: alarmId, idempotencyKey: idempotencyKey)] = result
    UserDefaults.standard.set(resolutions, forKey: resolutionsKey)
  }

  private static func storedOccurrences() -> [String: [String: Any]] {
    UserDefaults.standard.dictionary(forKey: occurrencesKey) as? [String: [String: Any]] ?? [:]
  }

  private static func storedResolutions() -> [String: [String: Any]] {
    UserDefaults.standard.dictionary(forKey: resolutionsKey) as? [String: [String: Any]] ?? [:]
  }

  private static func resolutionKey(alarmId: String, idempotencyKey: String) -> String {
    "\(alarmId):\(idempotencyKey)"
  }

  private static func scheduledFor(_ occurrence: [String: Any]) -> Int64 {
    (occurrence["scheduledFor"] as? NSNumber)?.int64Value ?? 0
  }
}

File: android/src/main/java/expo/modules/alarm/AlarmSchedulerOccurrenceStore.kt
package expo.modules.alarm

import android.content.Context
import org.json.JSONObject

internal object AlarmSchedulerOccurrenceStore {
  private const val PREFS_NAME = "alarm_scheduler_occurrences"
  private const val OCCURRENCE_PREFIX = "occurrence:"
  private const val RESOLUTION_PREFIX = "resolution:"

  private fun prefs(context: Context) =
    context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

  fun save(context: Context, occurrence: JSONObject) {
    val id = occurrence.optString("occurrenceId").takeIf(String::isNotBlank) ?: return
    prefs(context).edit().putString(OCCURRENCE_PREFIX + id, occurrence.toString()).apply()
  }

  fun occurrence(context: Context, occurrenceId: String): JSONObject? {
    val raw = prefs(context).getString(OCCURRENCE_PREFIX + occurrenceId, null) ?: return null
    return runCatching { JSONObject(raw) }.getOrNull()
  }

  fun all(context: Context, alarmId: String? = null): List<JSONObject> =
    prefs(context).all.mapNotNull { (key, value) ->
      if (!key.startsWith(OCCURRENCE_PREFIX)) return@mapNotNull null
      val occurrence = runCatching { JSONObject(value as? String ?: return@mapNotNull null) }.getOrNull()
        ?: return@mapNotNull null
      occurrence.takeIf { alarmId == null || it.optString("alarmId") == alarmId }
    }.sortedByDescending { it.optLong("scheduledFor") }

  fun updatePhase(context: Context, occurrenceId: String, phase: String) {
    val occurrence = occurrence(context, occurrenceId) ?: return
    occurrence.put("phase", phase)
    save(context, occurrence)
  }

  fun alarmId(context: Context, occurrenceId: String): String =
    occurrence(context, occurrenceId)?.optString("alarmId")?.takeIf(String::isNotBlank) ?: occurrenceId

  fun resolution(context: Context, alarmId: String, idempotencyKey: String): JSONObject? {
    val raw = prefs(context).getString(resolutionKey(alarmId, idempotencyKey), null) ?: return null
    return runCatching { JSONObject(raw) }.getOrNull()
  }

  fun saveResolution(context: Context, alarmId: String, idempotencyKey: String, result: JSONObject) {
    prefs(context).edit().putString(resolutionKey(alarmId, idempotencyKey), result.toString()).apply()
  }

  private fun resolutionKey(alarmId: String, idempotencyKey: String) =
    RESOLUTION_PREFIX + alarmId + ":" + idempotencyKey
}
